src/variableProcessor.ts

```diff
@@ -40,9 +79,130 @@ export class VariableProcessor {
             }
         }
 
+        // promise-based variables using URL-inspection must be last
+        let aadRegex = new RegExp(aadRegexPattern, 'g');
+        if (aadRegex.test(request)) {
+            request = request.replace(aadRegex, await VariableProcessor.getAadToken(request));
+        }
+
         return request;
     }
 
+    public static async getAadToken(url: string): Promise<string> {
+        // get target app from URL (fall back to empty string to support clear-only scenario)
+        let targetApp = (new RegExp("^[^\\s]+\\s+([^:]*:///?[^/]*/)").exec(url) || [url])[1] || "";
+
+        // detect known cloud URLs + fix audiences
+        let defaultCloud = null;
+        let cloud = null;
+        for (let c in Constants.AzureClouds) {
+            // set first cloud to be default
+            if (!defaultCloud) { defaultCloud = c; }
+
+            let found = false;
+            for (let api in Constants.AzureClouds[c]) {
+                if (api.endsWith("Audience")) { continue; }
+                if (targetApp === Constants.AzureClouds[c][api]) {
+                    cloud = c;
+                    targetApp = Constants.AzureClouds[c][api + "Audience"] || targetApp;
+                    found = true;
+                    break;
+                }
+            }
+            if (found) { break; }
+        }
+
+        // fall back to URL TLD
+        if (cloud === null) {
+            cloud = targetApp.substring(targetApp.lastIndexOf(".") + 1, targetApp.length - 1);
+        }
+
+        // parse input options -- [new|clear] [public|cn|de|us|ppe] [<domain|tenantId>]
+        let tenantId = "common";
+        let forceNewToken = false;
+        const groups = new RegExp(aadRegexPattern).exec(url);
+        if (groups) {
+            forceNewToken = (groups.length > 2 && groups[2] === Constants.AzureActiveDirectoryForceNewOption);

```

Comment:
no need to check groups array length here, it will safely return undefined if out of the array range

Reply:
Yeah, I saw there was a placeholder for every regex lookup, but I guess I was opting for safer code. Given we don't have a cross-browser situation here, you're right that this isn't necessary. I'll remove it.

Reply:
@flanakin since vscode extension run with Electron, which uses Chrome V8 internally. I think it's safe. 😄